app/grandchallenge/algorithms/views.py

```diff
@@ -342,41 +393,130 @@ def get_success_url(self):
+        return kwargs
 
-        if jobs["oldest"]:
-            next_job_at = jobs["oldest"] + period
-        else:
-            next_job_at = now
+    def get_context_data(self, *args, **kwargs):
+        context = super().get_context_data(*args, **kwargs)
+        context.update({"algorithm": self.algorithm})
+        context.update(
+            self.get_remaining_jobs(
+                credits_per_job=self.algorithm.credits_per_job
+            )
+        )
+        return context
 
-        if jobs["total"]:
-            total_jobs = user_credit.credits - jobs["total"]
-        else:
-            total_jobs = user_credit.credits
+    def form_valid(self, form):
+        def create_upload(image_files):
+            raw_files = []
+            upload_session = RawImageUploadSession.objects.create()
+            for image_file in image_files:
+                raw_files.append(
+                    RawImageFile(
+                        upload_session=upload_session,
+                        filename=image_file.name,
+                        staged_file_id=image_file.uuid,
+                    )
+                )
+            RawImageFile.objects.bulk_create(list(raw_files))
+            upload_session.save()
+            return upload_session.pk
+
+        civs = []
+        upload_pks = {}
+        for slug, value in form.cleaned_data.items():
+            ci = self.algorithm.inputs.get(slug=slug)
+            if ci.kind in InterfaceKind.interface_type_image():
+                # create civ without image, image will be added when import completes
+                civ = ComponentInterfaceValue.objects.create(interface=ci)
+                civ.save()
+                civs.append(civ.pk)
+                upload_pks[civ.pk] = create_upload(value)
+            elif ci.kind in InterfaceKind.interface_type_file():
+                # should be a single file
+                civ = ComponentInterfaceValue.objects.create(interface=ci)
+                civ.file = File(
+                    value[0].open(), name=str(Path(ci.output_path).name)
+                )
+                civ.full_clean()
+                civ.save()
+                civs.append(civ.pk)
+            else:
+                civ = ComponentInterfaceValue.objects.create(
+                    interface=ci, value=value
+                )
+                civs.append(civ.pk)
+
+        create_job = create_algorithm_job_for_inputs.signature(
+            kwargs={
+                "algorithm_image_pk": self.algorithm.latest_ready_image.pk,
+                "upload_pks": upload_pks,
+                "civ_pks": civs,
+                "creator_pk": self.request.user.pk,
+            },
+            immutable=True,
+        )
 
-        return {
-            "remaining_jobs": int(total_jobs / max(credits_per_job, 1)),
-            "next_job_at": next_job_at,
-            "user_credits": total_jobs,
-        }
+        self.job = create_job()
```

Comment:
We want to run this in place and not async? Is so, why is create_algorithm_job_for_inputs a shared celery task that needs to make database queries again for all the civs and algorithm image, and not just a method?

Reply:
Ah, it was made in place to get the job, so we can use it in the get_success_url. But you are right, might as well make it a method.

Reply:
Just check that this is possible, as what that method does might be slow...

Reply:
Also, you do not need to use get_success_url, you can always return your own HTTP redirect after calling `super().form_valid` (check the super method to see what that does).

Reply:
you mean in stead of calling `super()form_valid` right?